Fix ICE-133117: multiple never-pattern arm doesn't have false_edge_start_block

Shunpoco opened the pull request:
Fixes #133117 , and close fixes #133063 , fixes #130779 

In order to fix ICE-133117, at first I needed to tackle to ICE-133063 (this fixed 130779 as well).

### ICE-133063 and ICE-130779
This ICE is caused by those steps:
1. An arm has or-pattern, and all of the sub-candidates are never-pattern
2. In that case, all sub-candidates are removed in remove_never_subcandidates(). So the arm (candidate) has no sub-candidate.
3. In the current implementation, if there is no sub-candidate, the function assigns `pre_binding_block` into the candidate ([here](https://github.com/rust-lang/rust/blob/master/compiler/rustc_mir_build/src/builder/matches/mod.rs#L2002-L2004)). However, otherwise_block should be assigned to the candidate as well, because the otherwise_block is unwrapped in multiple place (like in lower_match_tree()). As a result, it causes the panic.

I simply added the same block as pre_binding_block into otherwise_block, but I'm wondering if there is a better block to assign to otherwise_block (is it ok to assign the same block into pre_binding and otherwise?)

### ICE-133117
This is caused by those steps:
1. There are two arms, both are or-pattern and each has one match-pair (in the test code, both are `(!|!)`), and the second arm has a guard.
2. In match_candidate() for the first arm, it expands the second arm’s sub-candidates as well ([here](https://github.com/rust-lang/rust/blob/master/compiler/rustc_mir_build/src/builder/matches/mod.rs#L1800-L1805)). As a result, the root candidate of the second arm is not evaluated/modified in match_candidate(). So a false_edge_start_block is not assigned to the candidate.
3. merge_trivial_subcandidates() is called against the candidate for the second arm. It just returns immediately because the candidate has a guard. So a flase_edge_start_block is not assigned to the candidate also in this function.
4. remove_never_subcandidates() is called against the candidate. Since all sub-candidates are never-pattern. they are removed.
5. In lower_match_tree(), since there is no sub-candidate for the candidate, the candidate itself is evaluated in visit_leave_rev ([here](https://github.com/rust-lang/rust/blob/master/compiler/rustc_mir_build/src/builder/matches/mod.rs#L1532)). Because the candidate has no false_edge_start_block, it causes the panic.

So I modified the order of if blocks in merge_trivial_subcandidates() to assign a false_edge_start_block if the candidate doesn't have.

Shunpoco committed:
Fix ICE-133063

If all subcandidates have never-pattern, the parent candidate should have otherwise_block
because some methods expect the candidate has the block.

Signed-off-by: Shunpoco <tkngsnsk313320@gmail.com>

Shunpoco committed:
Fix ICE-133117

If all subcandidates have never-pattern, we should assign false_edge_start_block to the parent candidate
if it doesn't have. merge_trivial_subcandidates does so, but if the candidate has guard it returns before the assignment.

Signed-off-by: Shunpoco <tkngsnsk313320@gmail.com>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12734524089/job/35492189066) [(plain)](https://github.com/rust-lang/rust/commit/8a57fa634c46273dab0a283d2ee151735342c514/checks/35492189066/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 27.2s done
#21 DONE 29.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
##[group]Testing stage2 compiletest suite=crashes mode=crashes (x86_64-unknown-linux-gnu)

running 234 tests
........................................................................................  88/234
.......................................................2025-01-12T15:08:12.577301Z ERROR compiletest::runtest: fatal error, panic: "crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge. If you want to see verbose output, set `COMPILETEST_VERBOSE_CRASHES=1`."
[crashes] tests/crashes/130779.rs ... F
[crashes] tests/crashes/130779.rs ... F
..............................2025-01-12T15:08:12.941936Z ERROR compiletest::runtest: fatal error, panic: "crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge. If you want to see verbose output, set `COMPILETEST_VERBOSE_CRASHES=1`."
[crashes] tests/crashes/133063.rs ... F
. 176/234
. 176/234
....2025-01-12T15:08:12.994236Z ERROR compiletest::runtest: fatal error, panic: "crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge. If you want to see verbose output, set `COMPILETEST_VERBOSE_CRASHES=1`."
[crashes] tests/crashes/133117.rs ... F
.....................................................

failures:
failures:

---- [crashes] tests/crashes/130779.rs stdout ----

error: crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge. If you want to see verbose output, set `COMPILETEST_VERBOSE_CRASHES=1`.
thread '[crashes] tests/crashes/130779.rs' panicked at src/tools/compiletest/src/runtest/crashes.rs:17:18:
fatal error
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


---- [crashes] tests/crashes/133063.rs stdout ----

error: crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge. If you want to see verbose output, set `COMPILETEST_VERBOSE_CRASHES=1`.
thread '[crashes] tests/crashes/133063.rs' panicked at src/tools/compiletest/src/runtest/crashes.rs:17:18:
fatal error

---- [crashes] tests/crashes/133117.rs stdout ----
---- [crashes] tests/crashes/133117.rs stdout ----

error: crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge. If you want to see verbose output, set `COMPILETEST_VERBOSE_CRASHES=1`.
thread '[crashes] tests/crashes/133117.rs' panicked at src/tools/compiletest/src/runtest/crashes.rs:17:18:
fatal error


```

</details>

Shunpoco committed:
Remove solved crashes

Signed-off-by: Shunpoco <tkngsnsk313320@gmail.com>

Shunpoco committed:
Fix ICE-130779

Signed-off-by: Shunpoco <tkngsnsk313320@gmail.com>

compiler-errors commented:
cc @Nadrieril

Nadrieril commented:
r? Nadrieril, thanks for doing this I'll have a look when I get a moment

Nadrieril commented on tests/ui/rfcs/rfc-0000-never_patterns/ICE-133063-never-arm-no-otherwise-block.rs:
```diff
@@ -0,0 +1,21 @@
+    None
+}
+
+fn assign_twice() {
+    loop {
+        match () {
+            (!| //~ ERROR: mismatched types
+            !) if let _ = split_last(&()) => {} //~ ERROR a never pattern is always unreachable
+            //~^ ERROR: mismatched types
+            //~^^ WARNING: irrefutable `if let` guard pattern [irrefutable_let_patterns]
+            _ => {}
+        }
+    }
+}
```
Could you replace this with a type-correct test, so this test keeps being useful in the future if the implementation changes? In particular we may skip MIR lowering in case of a type error, which would render the current test useless.
```rust
fn foo(x: !) {
	match x {
		(! | !) if false => {}
		_ => {}
	}
}
```

Shunpoco replied:
Addressed in https://github.com/rust-lang/rust/pull/135409/commits/d8a216b866ff238589005fb1ea4c54644d8d24bf.

Note: `x: !` can't avoid mismatched types error. So I use enum Void {} instead of it (other test codes in the folder do so).

Nadrieril commented on tests/ui/rfcs/rfc-0000-never_patterns/ICE-133117-duplicate-never-arm.rs:
```diff
@@ -0,0 +1,12 @@
+#![feature(never_patterns)]
+#![allow(incomplete_features)]
+
+fn main() {
+    match () {
+        (!|
+        //~^ ERROR: mismatched types
+        !) if true => {} //~ ERROR a never pattern is always unreachable
+        //~^ ERROR: mismatched types
+        (!|!) if true => {} //~ ERROR a never pattern is always unreachable
+    }
+}
```
Same here, could you make the test type-correct:
```rust
#![feature(never_type)]
#![feature(never_patterns)]
#![allow(incomplete_features)]

fn foo(x: !) {
    match x {
        (!|!) if true => {}
        (!|!) if true => {}
    }
}
```

Shunpoco replied:
Fixed in https://github.com/rust-lang/rust/pull/135409/commits/e7db0925d1441990ceb8236b5d4015f9a2ba333c.

Nadrieril commented on compiler/rustc_mir_build/src/builder/matches/mod.rs:
```diff
@@ -1940,6 +1940,10 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
     /// in match tree lowering.
     fn merge_trivial_subcandidates(&mut self, candidate: &mut Candidate<'_, 'tcx>) {
         assert!(!candidate.subcandidates.is_empty());
+        if candidate.false_edge_start_block.is_none() {
+            candidate.false_edge_start_block = candidate.subcandidates[0].false_edge_start_block;
+        }
+
```
I'd rather not do this here; the invariant is that candidates with subcandidates shouldn't have `pre_binding`/`otherwise`/`false_edge_start` blocks, so we should only assign that after merge. This should be assigned in `remove_never_subcandidates` instead I think.

Shunpoco replied:
Thanks. I modified it in https://github.com/rust-lang/rust/pull/135409/commits/481ed2d9319a28b770da60f4567e55f5cc053835.

Shunpoco committed:
address review: modify ICE-133117-duplicated-never-arm.rs

Use enum Void to avoid mistmatched types error

Signed-off-by: Shunpoco <tkngsnsk313320@gmail.com>

Shunpoco committed:
address review: modify ICE-133063-never-arm-no-otherwise-block.rs

- Use enum Void to avoid mismatched types error
- We don't need to use if let to check the ICE

Signed-off-by: Shunpoco <tkngsnsk313320@gmail.com>

Shunpoco committed:
address review: modify matches/mod.rs

The candidate shouldn't have false_edge_start_block if it has sub candidates.
In remove_never_subcandidates(), the false_edge_start_block from the first sub candidte is assigned to a value and the value is later used if all sub candidates are removed and candidate doesn't have false_edge_start_block.
In merge_trivial_subcandidates, I leave the if block which assign a false_edge_start_block into the candidate as before I put this commit since compile panics.

Signed-off-by: Shunpoco <tkngsnsk313320@gmail.com>

Shunpoco committed:
modify comment

Signed-off-by: Shunpoco <tkngsnsk313320@gmail.com>

Shunpoco commented:
@rustbot review

Nadrieril commented:
Looks good, thank you for investigating and fixing this :)

@bors r+

Nadrieril commented:
@bors rollup
